File: Binary-Trees/increasing-order-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorder(root *TreeNode, vals []int) []int {
    if root == nil {
        return vals
    }
     
    vals = inorder(root.Left,vals)
    vals = append(vals,root.Val)
    vals = inorder(root.Right,vals)
    
    return vals
}

func increasingBST(root *TreeNode) *TreeNode {
    
    if root==nil {
        return nil
    }
    
    vals := inorder(root,[]int{})
    hdnode := &TreeNode{Val:vals[0]}
    slnode := hdnode
    for _,val := range vals[1:] {
        newNode := &TreeNode{Val:val}
        slnode.Right = newNode
        slnode = slnode.Right
    }
    
    return hdnode
}

File: Binary-Trees/increasing-order-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    TreeNode cur = new TreeNode(0);
    public TreeNode increasingBST(TreeNode root) {
        TreeNode ans = new TreeNode(0);
        ans = cur;
        inorder(root);
        return ans.right;
    }
    
    public void inorder(TreeNode root)
    {
        if(root==null)
            return;
        inorder(root.left);
        root.left = null;
        cur.right = root;
        cur = root;
        inorder(root.right);
    }
}
